File: core/src/main/javascript/userauth/UserAuthenticationData.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * <p>The user authentication data provides proof of user identity.</p>
 *
 * <p>Specific user authentication mechanisms should define their own user
 * authentication data types.</p>
 *
 * <p>User authentication data is represented as
 * {@code
 * userauthdata = {
 *   "#mandatory" : [ "scheme"., "authdata" ],
 *   "scheme" : "string",
 *   "authdata" : object
 * }} where
 * <ul>
 * <li>{@code scheme} is the user authentication scheme</li>
 * <li>{@code authdata} is the scheme-specific authentication data</li>
 * </ul></p>
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
	"use strict";
	
	var MslEncodable = require('../io/MslEncodable.js');
	var AsyncExecutor = require('../util/AsyncExecutor.js');
	var MslUserAuthException = require('../MslUserAuthException.js');
	var MslError = require('../MslError.js');
	var MslEncoderException = require('../io/MslEncoderException.js');
	var MslEncodingException = require('../MslEncodingException.js');
	
    /**
     * Key user authentication scheme.
     * @const
     * @type {string}
     */
    var KEY_SCHEME = "scheme";
    /**
     * Key user authentication data.
     * @const
     * @type {string}
     */
    var KEY_AUTHDATA = "authdata";

    var UserAuthenticationData = module.exports = MslEncodable.extend({
        /**
         * Create a new user authentication data object with the specified user
         * authentication scheme.
         *
         * @param {UserAuthenticationScheme} scheme the user authentication scheme.
         * @constructor
         * @interface
         */
        init: function init(scheme) {
            // The properties.
            var props = {
                /**
                 * User authentication scheme.
                 * @type {UserAuthenticationScheme}
                 */
                scheme: { value: scheme, writable: false, configurable: false },
                /**
                 * Cached encodings.
                 * @type {Object<MslEncoderFormat,Uint8Array>}
                 */
                encodings: { value: {}, writable: false, enumerable: false, configurable: false },
            };
            Object.defineProperties(this, props);
        },

        /**
         * Returns the scheme-specific user authentication data. This method is
         * expected to succeed unless there is an internal error.
         * 
         * @param {MslEncoderFactory} encoder the encoder factory.
         * @param {MslEncoderFormat} format the encoder format.
         * @param {{result: function(MslObject}, error: function(Error)}
         *        callback the callback that will receive the authentication
         *        data MSL object or any thrown exceptions.
         * @throws MslEncoderException if there was an error constructing the
         *         MSL object.
         */
        getAuthData: function(encoder, format, callback) {},

        /**
         * @param {Object} that the object with which to compare.
         * @return {boolean} true if this object is equal to that object.
         */
        equals: function equals(that) {
            if (this === that) return true;
            if (!(that instanceof UserAuthenticationData)) return false;
            return this.scheme == that.scheme;
        },
        
        /** @inheritDoc */
        toMslEncoding: function toMslEncoding(encoder, format, callback) {
            var self = this;
            AsyncExecutor(callback, function() {
                // Return any cached encoding.
                if (this.encodings[format])
                    return this.encodings[format];
                
                this.getAuthData(encoder, format, {
                    result: function(authdata) {
                        AsyncExecutor(callback, function() {
                            // Encode the user authentication data.
                            var mo = encoder.createObject();
                            mo.put(KEY_SCHEME, this.scheme.name);
                            mo.put(KEY_AUTHDATA, authdata);
                            encoder.encodeObject(mo, format, {
                            	result: function(encoding) {
                            		AsyncExecutor(callback, function() {
                            			// Cache and return the encoding.
                            			this.encodings[format] = encoding;
                            			return encoding;
                            		}, self);
                            	},
                            	error: callback.error
                            });
                        }, self);
                    },
                    error: callback.error,
                });
            }, self);
        },
    });

    /**
     * <p>Construct a new user authentication data instance of the correct type
     * from the provided MSL object.</p>
     * 
     * <p>A master token may be required for certain user authentication
     * schemes.</p>
     *
     * @param {MslContext} ctx MSL context.
     * @param {MasterToken} masterToken the master token associated with the user
     *        authentication data. May be {@code null}.
     * @param {MslObject} userAuthMo the MSL object.
     * @param {{result: function(UserAuthenticationData), error: function(Error)}}
     *        callback the callback functions that will receive the user
     *        authentication data or any thrown exceptions.
     * @return {UserAuthenticationData} the user authentication data concrete instance.
     * @throws MslEncodingException if there is an error parsing the data.
     * @throws MslUserAuthException if there is an error instantiating the user
     *         authentication data.
     * @throws MslCryptoException if there is an error with the entity
     *         authentication data cryptography.
     */
    var UserAuthenticationData$parse = function UserAuthenticationData$parse(ctx, masterToken, userAuthMo, callback) {
        AsyncExecutor(callback, function() {
            try {
                // Pull the scheme.
                var schemeName = userAuthMo.getString(KEY_SCHEME);
                var scheme = ctx.getUserAuthenticationScheme(schemeName);
                if (!scheme)
                    throw new MslUserAuthException(MslError.UNIDENTIFIED_USERAUTH_SCHEME, schemeName);

                // Construct an instance of the concrete subclass.
                var factory = ctx.getUserAuthenticationFactory(scheme);
                if (!factory)
                    throw new MslUserAuthException(MslError.USERAUTH_FACTORY_NOT_FOUND, scheme.name);
                var encoder = ctx.getMslEncoderFactory();
                factory.createData(ctx, masterToken, userAuthMo.getMslObject(KEY_AUTHDATA, encoder), callback);
            } catch (e) {
                if (e instanceof MslEncoderException)
                    throw new MslEncodingException(MslError.MSL_PARSE_ERROR, "userauthdata " + userAuthMo, e);
                throw e;
            }
        });
    };
    
    // Exports.
    module.exports.parse = UserAuthenticationData$parse;
})(require, (typeof module !== 'undefined') ? module : mkmodule('UserAuthenticationData'));

File: core/src/main/cpp/userauth/UserAuthenticationData.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <io/MslEncoderFactory.h>
#include <io/MslObject.h>
#include <MslEncodingException.h>
#include <MslError.h>
#include <MslUserAuthException.h>
#include <tokens/MasterToken.h>
#include <userauth/UserAuthenticationData.h>
#include <userauth/UserAuthenticationFactory.h>
#include <util/MslContext.h>
#include <util/MslUtils.h>

using namespace std;
using namespace netflix::msl;
using namespace netflix::msl::io;
using namespace netflix::msl::tokens;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace userauth {

namespace {

/** Key user authentication scheme. */
const string KEY_SCHEME = "scheme";
/** Key user authentication data. */
const string KEY_AUTHDATA = "authdata";

} // namespace anonymous

// static
shared_ptr<UserAuthenticationData> UserAuthenticationData::create(shared_ptr<MslContext> ctx,
        shared_ptr<MasterToken> masterToken,
        shared_ptr<MslObject> userAuthMo)
{
    try {
        // Pull the scheme.
        const string schemeName = userAuthMo->getString(KEY_SCHEME);
        UserAuthenticationScheme scheme = ctx->getUserAuthenticationScheme(schemeName);
        if (scheme == UserAuthenticationScheme::INVALID)
            throw MslUserAuthException(MslError::UNIDENTIFIED_USERAUTH_SCHEME, schemeName);

        // Construct an instance of the concrete subclass.
        shared_ptr<UserAuthenticationFactory> factory = ctx->getUserAuthenticationFactory(scheme);
        if (!factory)
            throw MslUserAuthException(MslError::USERAUTH_FACTORY_NOT_FOUND, scheme.name());
        shared_ptr<MslEncoderFactory> encoder = ctx->getMslEncoderFactory();
        return factory->createData(ctx, masterToken, userAuthMo->getMslObject(KEY_AUTHDATA, encoder));
    } catch (const MslEncoderException& e) {
        throw MslEncodingException(MslError::MSL_PARSE_ERROR, "userauthdata " + userAuthMo->toString(), e);
    }
}

shared_ptr<ByteArray> UserAuthenticationData::toMslEncoding(shared_ptr<MslEncoderFactory> encoder,
        const MslEncoderFormat& format) const
{
    // Return any cached encoding.
    map<MslEncoderFormat, shared_ptr<ByteArray>>::const_iterator it = encodings_.find(format);
    if (it != encodings_.end())
        return it->second;

    // Encode the user authentication data.
    shared_ptr<MslObject> mo = encoder->createObject();
    mo->put(KEY_SCHEME, scheme_.name());
    mo->put<shared_ptr<MslObject>>(KEY_AUTHDATA, getAuthData(encoder, format));
    shared_ptr<ByteArray> encoding = encoder->encodeObject(mo, format);

    // Cache and return the encoding.
    encodings_.insert(make_pair(format, encoding));
    return encoding;
}

bool UserAuthenticationData::equals(shared_ptr<const UserAuthenticationData> that) const
{
    if (!that) return false;
    if (this == that.get()) return true;
    return scheme_ == that->scheme_;
}

bool operator==(const UserAuthenticationData& a, const UserAuthenticationData& b)
{
	shared_ptr<const UserAuthenticationData> ap(&a, &MslUtils::nullDeleter<UserAuthenticationData>);
	shared_ptr<const UserAuthenticationData> bp(&b, &MslUtils::nullDeleter<UserAuthenticationData>);
	return ap->equals(bp);
}


}}} // namespace netflix::msl::userauth
